Wojciech Matejuk: easy part

WojciechMat opened the pull request:
I have decided to create my first pull request after finishing first stage of the project for you to have more time to review it and for me to get the feedback as soon as possible.
### Speed
**1.** Example of a chart created with speed.py (Notes per minute):

<a href="url"><img src="https://github.com/Nospoko/midi-internship/assets/74838859/c828096b-d163-4e68-a93d-22c16705e439" align="centre" height="300" width="500" ></a>
<br>
**2.** Number of notes pressed simultaneously:
The number on the beginning of a title says what time threshold was used in seconds.

<a href="url"><img src="https://github.com/Nospoko/midi-internship/assets/74838859/ba5ecb78-6a28-41ef-b909-b16433971b0b" align="left" height="240" width="380" ></a> <a href="url"><img src="https://github.com/Nospoko/midi-internship/assets/74838859/25d303bd-9546-4d0f-b0e0-66c2cd793796" align="right" height="240" width="380" ></a>
<br>
<a href="url"><img src="https://github.com/Nospoko/midi-internship/assets/74838859/ddbca95e-ef54-46d5-b297-38bff29f89a6" align="left" height="240" width="380" ></a> <a href="url"><img src="https://github.com/Nospoko/midi-internship/assets/74838859/baf4a3bd-c94a-4b38-b945-2c3dac8aeb37" align="right" height="240" width="380" ></a>
<br>
The charts show the number of notes pressed simultaneously within different time thresholds. It's interesting to note that the number of simultaneous notes does not always increase with a larger time threshold, which suggests the need for a different grouping method to achieve more accuracy.

Using functions I developed while creating this program, I processed the 'roszcz/maestro-v1' database and found that the piece with the fastest 15 seconds of music is 'Etude Op. 25 No. 10 in B Minor' by Frédéric Chopin, containing 483 notes in the fastest 15-second timeframe. The algorithm took approximately 32 seconds to search through the database and create a CSV file containing the info on the fastest 15-second timeframe for each file.

### Chords
Chords are recognized from groups of notes playing together.

**1.** Examples of a chart created with chords.py:

<a href="url"><img src="https://github.com/Nospoko/midi-internship/assets/74838859/56267200-bc58-4b41-9ac5-4b2249eae271" align="centre" height="300" width="500" ></a>
<br>
<a href="url"><img src="https://github.com/Nospoko/midi-internship/assets/74838859/727c0b7f-f745-462b-889c-bd0811de741e" align="centre" height="300" width="500" ></a>
<br>
**2.** I created a table of most commonly repeated chord in each of  the pieces. Please note that a single chart for each piece can be created in a future pull request.
full version: [most-played-chords.csv](https://github.com/Nospoko/midi-internship/files/12210237/most-played-chords.csv)
|id    |composer                   |title                                       |chord |repetitions|
|------|---------------------------|--------------------------------------------|------|-----------|
|0     |Alban Berg                 |Sonata Op. 1                                |DM    |71         |
|1     |Alban Berg                 |Sonata Op. 1                                |F#M   |68         |
|2     |Alban Berg                 |Sonata Op. 1                                |F#M   |58         |
|...   |...                        |...                                         |...   |...        |
|99    |Johannes Brahms            |Piano Concerto No. 1 in D Minor, Op. 15     |AM    |57         |
|100   |Johannes Brahms            |Piano Sonata No. 2 in F-Sharp Minor, Op. 2  |F#M   |72         |
|101   |Johannes Brahms            |Rhapsody in B Minor, Op. 79, No. 1          |F#M   |90         |
|...   |...                        |...                                         |...   |...        |

The implementation could be optimized due to several conversions. Iterating through the entire database took approximately 25 minutes. The algorithm concluded that the piece with the highest number of repetitions of one chord is:
Composer: Johannes Brahms
Title: Variations on a Theme by Paganini, Op. 35, Volumes 1 & 2
Chord: EM
Repetitions: 783

Please let me know if you have any feedback or suggestions for improvements.

roszcz commented:
Thanks @WojciechMat, that's great!

I'm having some issues running your `chords.py` file:

```
  File "/home/hagrid/workson/midi-internship/src/chords.py", line 54, in MidiPiece_to_MidiFile
    mido_obj = MidiFile("res.mid", ticks_per_beat=480)
  File "/home/hagrid/.virtualenvs/midi-internship/lib/python3.9/site-packages/miditoolkit/midi/parser.py", line 66, in __init__
    self.instruments = self._load_instruments(mido_obj)
  File "/home/hagrid/.virtualenvs/midi-internship/lib/python3.9/site-packages/miditoolkit/midi/parser.py", line 205, in _load_instruments
    current_instrument = np.zeros(16, dtype=np.int)
  File "/home/hagrid/.virtualenvs/midi-internship/lib/python3.9/site-packages/numpy/__init__.py", line 313, in __getattr__
    raise AttributeError(__former_attrs__[attr])
AttributeError: module 'numpy' has no attribute 'int'.
`np.int` was a deprecated alias for the builtin `int`. To avoid this error in existing code, use `int` by itself. Doing this will not modify any behavior and is safe. When replacing `np.int`, you may wish to use e.g. `np.int64` or `np.int32` to specify the precision. If you wish to review your current use, check the release note link for additional information.
The aliases was originally deprecated in NumPy 1.20; for more details and guidance see the original release note at:
    https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations
```

It looks like it's related to the `miditoolkit` library not being up to date with numpy. Are you using a <1.20 version of numpy locally? 

Another way around this problem would be not to do the `MidiPiece -> PrettyMIDI -> miditoolkit` conversion, and create the list of Note objects directly:

```python
from miditoolkit.midi.containers import Note


def piece_to_notes(piece: ff.MidiPiece) -> list[Note]:
    notes = []
    columns = ["start", "end", "pitch", "velocity"]
    for it, row in piece.df[columns].iterrows():
        note = Note(**row.to_dict())
        notes.append(note)

    return notes
```

But I see that you're also doing some time conversion, so there would be more changes required.

WojciechMat commented:
@roszcz Thank you for the feedback!

Actually, I have changed miditoolkit/midi/parser.py file in my  miditoolkit library to use int instead of np.int, so it could be used with numpy >1.20, you can do the same if you want to use newest numpy version. Or you can wait for me to update the code to delete unnecessary conversions.

I used unnecessary conversions because I was getting to know chorder library and it’s possibilities. I will use a better way of parsing notes soon.

roszcz commented:
It's interesting to see the discrepancies in results for the same piece in the csv you shared:

```
Franz Schubert	Sonata in A Min.	CM	143
Franz Schubert	Sonata in A Min.	CM	260
Franz Schubert	Sonata in A Min.	EM	135
Franz Schubert	Sonata in A Min.	EM	189
```

For the same piece, I think we should expect the number of chord to be very close. 

The detected chord name is also interesting - I'm guessing that maybe there's some issue with Major and Minor chord names - CMajor7 (C E G H) could be easily confused with EMinor (E G H), but not with EMajor (E G# H). From reading your code, it looks like this is the full chorder output, so I'm not sure where is the confusion coming from. 

I'll try to take a more detailed look later today :)

WojciechMat commented:
@roszcz I have listened to those four pieces saved as Sonata in A Min (position 280, 281, 282 and 283 in database["train"+"test"+"validation"], and it seems like these are totally different pieces, so that is probably where the discrepances come from.
I have also removed unnecessary to and from file conversion from midipiece_to_midifile function and it now takes only 15 minutes to iterate through the database. I will commit the changes in the evening and will try to see if the database is wrong or if I am naming the pieces incorectlly.

roszcz commented:
This sort of makes sense - "sonata" is a musical structure consisting of 3 parts, that are often performed separately, which could explain those differences. It's also great to know that there's this inconsistency in the maestro dataset, thanks!

WojciechMat commented:
@roszcz I have updated Pull Request description and am waiting for your feedback. I will use this stage in the task of finding similarities. Have a great day.

Ps. I plan on adding visual representation of files in the future.

roszcz commented on src/chords.py:
```diff
@@ -0,0 +1,164 @@
+    )
+
+
+def process_dataset(dataset, csv_file_path):
```
```suggestion
def process_dataset(dataset: Dataset) -> pd.DataFrame:
```

I think it's a good practice to separate data processing logic from any data presentation/visualization procedures, so the csv creation could be moved to the main part:

```python
df = process_dataset(dataset)
df.to_csv(csv_file_path)
```

I also included type hints in my suggestion - these are very helpful in maintaining readability :)

roszcz commented on src/chords.py:
```diff
@@ -0,0 +1,164 @@
+        # Pop notes which ended before current note started
+        if pq.empty():
+            continue
+        while note.start > pq.queue[0].end:
```
I'm not sure if I'm reading this correctly, but it looks like as long as the first note of the sequence is being hold, everything after it is going to be considered part of the chord.

For example, this gets recognized as CM:

```
   pitch  start  end  duration  velocity
0     60      0    3         3        80
1     64      1    3         2        80
2     67      2    3         1        80
```

But there is one second distance between each of the notes, so the expected output would be "no chords".

The original intention was to find chords defined as notes performed simultaneously - the issue here being that in reality those types of chords are spread over (short amount of) time - either due to aesthetic choices of the player, or just because it's almost impossible to press multiple keys at the exactly same time with millisecond precision :)

Your logic can be adjusted for those requirements with some kind of thresholding for the maximum allowed duration of the chord:

```python
while pq.queue[0].end - note.start  < threshold:
```

roszcz commented on src/chords.py:
```diff
@@ -0,0 +1,164 @@
+    count = 0
+    chords = []
+    prev_chord = None
+    midi_data = piece.df
+    piece = piece.to_midi()
+    midi_data["start"] = midi_data.apply(start_to_ticks, axis=1, args=(piece,))
+    midi_data["end"] = midi_data.apply(end_to_ticks, axis=1, args=(piece,))
```
```suggestion
    midi_data = piece.df.copy()
    piece = piece.to_midi()
    midi_data["start"] = midi_data.apply(start_to_ticks, axis=1, args=(piece,))
    midi_data["end"] = midi_data.apply(end_to_ticks, axis=1, args=(piece,))
```

Your version modifies the internal dataframe of the piece, which could lead to serious confusion if the same object had to be used for any other processing after chord detection. Making a copy is a safe way to prevent those problems.

WojciechMat replied:
Thank you very much!

WojciechMat commented:
I have incorporated your suggestions and added functionality.
## Chords
### Chord recognition
I decided to use 24 ticks as a threshold for maximum chord duration. The results of finding most repeated chord in databes have changed and are more consistent with the results in [multiples.csv](https://github.com/Nospoko/midi-internship/files/12230909/multiples.csv)

### Most repeated chords
Here are new results of finding the most repeated chord for each piece:

[most-played-chords-v2.csv](https://github.com/Nospoko/midi-internship/files/12230929/most-played-chords-v2.csv)

The piece with the most repetitions is:
composer:                **Franz Schubert**
title:          **Sonata in D Major, D850**
chord:                               **DM**
repetitions                       **1582**

### Chord tables

Here is an example of chord table created for Edvard Erieg, Lyric piece in e minor, “waltz,” op 38 no 7:

<img src="https://github.com/Nospoko/midi-internship/assets/74838859/a2645f7e-280a-4af2-9850-53232e11f8d6" align="centre" height="800" width="380" ></a>
<br>

**csv files**:

[franz-liszt-dante-sonata.csv](https://github.com/Nospoko/midi-internship/files/12230646/franz-liszt-dante-sonata.csv)

[franz-schubert-sonata-in-d-major,-d850.csv](https://github.com/Nospoko/midi-internship/files/12230667/franz-schubert-sonata-in-d-major.-d850.csv)

## Finding inconsistent data - notes
I have updated similarity threshold. The algorithm does result in some false-positives (pieces that are the same are sometimes flagged as not similar ), but it can help in renaming the data.

**The method with Leveshtein distance does not recognize the same pieces in different keys as similar.**

**new csv files**:
[multiples.csv](https://github.com/Nospoko/midi-internship/files/12231006/multiples.csv)

[inconsistent-data.csv](https://github.com/Nospoko/midi-internship/files/12231004/inconsistent-data.csv)
